plugin/common/kernels/gridAnchorLayer.cu

```diff
@@ -19,105 +19,140 @@
+    float yMax = yC + 0.5 * heights[arId];
+
+    outputData[tid * 4] = xMin;
+    outputData[tid * 4 + 1] = yMin;
+    outputData[tid * 4 + 2] = xMax;
+    outputData[tid * 4 + 3] = yMax;
+
+    // Remember to move the output cursor
+    float* output = outputData + dim * 4;
+
+    // Simply copying the variance
+    output[tid * 4] = param.variance[0];
+    output[tid * 4 + 1] = param.variance[1];
+    output[tid * 4 + 2] = param.variance[2];
+    output[tid * 4 + 3] = param.variance[3];
 }
 
-namespace nvinfer1
-{
-namespace plugin
+using nvinfer1::plugin::reduced_divisor;
+template <unsigned nthdsPerCTA>
+__launch_bounds__(nthdsPerCTA)
+__global__ void gridAnchorRectKernel(
+    const GridAnchorParameters param,
+    const int numAspectRatios,
+    reduced_divisor divObj,
+    const float* widths,
+    const float* heights,
+    float* outputData
+    )
 {
+    // output dims: (H, W, param.numMinSize, (1+haveMaxSize+numAR-1), 4)
+    const int dim = param.H * param.W * numAspectRatios;
+
+    /*
+     * Parameters used to calculate the bounding box coordinates back to input image scale
+     * Normally we calculate the anchorStride = image_input_size (in pixel) / feature_map_size
+     * Here we do not use image_input_size for the moment
+     * Instead we use 1.0
+     * The coordinates calculated are scaled by the input image size.
+     * Most of the coordinates will be in a range of [0, 1], except for the bounding box coordinates going outside of the image
+     * Every coordinate will go back to the pixel coordinates in the input image if being multiplied by image_input_size
+     * Here we implicitly assumes the image input and feature map are square
+     */
+    float anchorStrideH = (1.0 / param.H);
+    float anchorStrideW = (1.0 / param.W);
+    float anchorOffsetH = 0.5 * anchorStrideH;
+    float anchorOffsetW = 0.5 * anchorStrideW;
+
+    int tid = blockIdx.x * blockDim.x + threadIdx.x;
+    if (tid >= dim)
+    {
+        return;
+    }
+
+    int arId, currIndex;
+    divObj.divmod(tid, currIndex, arId);
+
+    const int w = currIndex % param.W;
+    const int h = currIndex / param.W;
+
+    // Center coordinates
+    float yC = h * anchorStrideH + anchorOffsetH;
+    float xC = w * anchorStrideW + anchorOffsetW;
+
+    // x_min, y_min
+    float xMin = xC - 0.5 * widths[arId];
+    float yMin = yC - 0.5 * heights[arId];
+
+    // x_max, y_max
+    float xMax = xC + 0.5 * widths[arId];
+    float yMax = yC + 0.5 * heights[arId];
+
+    outputData[tid * 4] = xMin;
+    outputData[tid * 4 + 1] = yMin;
+    outputData[tid * 4 + 2] = xMax;
+    outputData[tid * 4 + 3] = yMax;
+
+    // Remember to move the output cursor
+    float* output = outputData + dim * 4;
+
+    // Simply copying the variance
+    output[tid * 4] = param.variance[0];
+    output[tid * 4 + 1] = param.variance[1];
+    output[tid * 4 + 2] = param.variance[2];
+    output[tid * 4 + 3] = param.variance[3];
+}
+
 pluginStatus_t anchorGridInference(
```

Comment:
Doesn't this function need to be update to call the new `gridAnchorRectKernel` kernel?

Reply:
Removed redundant kernel. Note: The kernel already supported rectangular feature maps.